### PR TITLE
Update crop insurance endpoints to use database

### PR DESCRIPTION
**Pull request by ywkim312**
To try this, use the database named `pdl_ywkim_test_ci` in dev server.

### state-distribution
The original endpoint example
```
{
  "2018-2022": [
    {
      "state": "TX",
      "programs": [
        {
          "programName": "Crop Insurance",
          "totalIndemnitiesInDollars": 10192200379,
          "totalPremiumInDollars": 6715854749,
          "totalPremiumSubsidyInDollars": 4527376715,
          "totalFarmerPaidPremiumInDollars": 2188478034,
          "totalNetFarmerBenefitInDollars": 8003722345,
          "totalPoliciesEarningPremium": 368335,
          "averageLiabilitiesInDollars": 5808509412.2,
          "averageInsuredAreaInAcres": 43828206.2,
          "lossRatio": 1.518,
          "subPrograms": []
        }
      ]
    }
```

The result in this PR
```
{
  "2018-2022": [
    [
      "TX",
      {
        "programName": "Crop Insurance",
        "totalIndemnitiesInDollars": 10192200379,
        "totalPremiumInDollars": 6715854749,
        "totalPremiumSubsidyInDollars": 4527376715,
        "totalFarmerPaidPremiumInDollars": 2188478034,
        "totalNetFarmerBenefitInDollars": 8003722345,
        "totalPoliciesEarningPremium": 368335,
        "averageLiabilitiesInDollars": 5808509412.2,
        "averageInsuredAreaInAcres": 43828206.2,
        "lossRatio": 1.518,
        "subPrograms": []
      }
    ],
```

### Summary
The original endpoint
```
{
  "programs": [
    {
      "programName": "Crop Insurance",
      "subPrograms": [],
      "totalIndemnitiesInDollars": 54219102283,
      "totalPremiumInDollars": 60851328009,
      "totalPremiumSubsidyInDollars": 38304200436,
      "totalFarmerPaidPremiumInDollars": 22547127573,
      "totalNetFarmerBenefitInDollars": 31671974710,
      "averageLiabilitiesInDollars": 2443164592.59,
      "averageInsuredAreaInAcres": 8201387.89,
      "totalPoliciesEarningPremium": 5637523,
      "lossRatio": 0.891
    }
  ]
}

```

This PR
```
{
  "programName": "Crop Insurance",
  "totalIndemnitiesInDollars": 54219102283,
  "totalPremiumInDollars": 60851328009,
  "totalPremiumSubsidyInDollars": 38304200436,
  "totalFarmerPaidPremiumInDollars": 22547127573,
  "totalNetFarmerBenefitInDollars": 31671974710,
  "totalPoliciesEarningPremium": 5637523,
  "averageLiabilitiesInDollars": 2443164592.59,
  "averageInsuredAreaInAcres": 8201387.89,
  "lossRatio": 0.891,
  "subPrograms": []
}
```

### CODE REVIEW

**Review by pengyin-shan — CHANGES_REQUESTED** [review]
For the updated CRP (and EQIP, CSP), we are now using the "state" attribute instead of the state name directly for `state-distribution`:
Screenshot 2024-09-19 at 5 31 42 PM
Additionally, we have removed the "programName" attribute in the summary section:
Screenshot 2024-09-19 at 5 32 27 PM
Could we ensure that the new crop insurance data follows the same structure?

**Comment by ywkim312** [comment]
> {
>   "2018-2022": [
>     [
>       "TX",
>       {
>         "programName": "Crop Insurance",
>         "totalIndemnitiesInDollars": 10192200379,
>         "totalPremiumInDollars": 6715854749,
>         "totalPremiumSubsidyInDollars": 4527376715,
>         "totalFarmerPaidPremiumInDollars": 2188478034,
>         "totalNetFarmerBenefitInDollars": 8003722345,
>         "totalPoliciesEarningPremium": 368335,
>         "averageLiabilitiesInDollars": 5808509412.2,
>         "averageInsuredAreaInAcres": 43828206.2,
>         "lossRatio": 1.518,
>         "subPrograms": []
>       }
>     ],

Updated both, so it looks like

```{
  "2018-2022": [
    {
      "state": "TX",
      "programName": "Crop Insurance",
      "totalIndemnitiesInDollars": 10192200379,
      "totalPremiumInDollars": 6715854749,
      "totalPremiumSubsidyInDollars": 4527376715,
      "totalFarmerPaidPremiumInDollars": 2188478034,
      "totalNetFarmerBenefitInDollars": 8003722345,
      "totalPoliciesEarningPremium": 368335,
      "averageLiabilitiesInDollars": 5808509412.2,
      "averageInsuredAreaInAcres": 43828206.2,
      "lossRatio": 1.518,
      "subPrograms": []
    },
    ```
    and
    
    ```
{
  "totalIndemnitiesInDollars": 54219102283,
  "totalPremiumInDollars": 60851328009,
  "totalPremiumSubsidyInDollars": 38304200436,
  "totalFarmerPaidPremiumInDollars": 22547127573,
  "totalNetFarmerBenefitInDollars": 31671974710,
  "totalPoliciesEarningPremium": 5637523,
  "averageLiabilitiesInDollars": 2443164592.59,
  "averageInsuredAreaInAcres": 8201387.89,
  "lossRatio": 0.891,
  "subPrograms": []
}
```

**Review by pengyin-shan — APPROVED** [review]
Great, thank you for the update! I’m approving the changes. Please let me know once this is on the dev environment, and I’ll proceed with the front-end update.

**Comment by ywkim312** [comment]
> Great, thank you for the update! I’m approving the changes. Please let me know once this is on the dev environment, and I’ll proceed with the front-end update.

It is little tricky. There are PR in the data-ingestion and these decide the content of the database. If we create a database for crop insurance, there will no ACEP or RCPP and vice-versa. I will update it when the PR get merged so the database has all the information.

**Comment by pengyin-shan** [comment]
> > Great, thank you for the update! I’m approving the changes. Please let me know once this is on the dev environment, and I’ll proceed with the front-end update.
> 
> It is little tricky. There are PR in the data-ingestion and these decide the content of the database. If we create a database for crop insurance, there will no ACEP or RCPP and vice-versa. I will update it when the PR get merged so the database has all the information.

Ah I see, please take your time. Thank you!

**Comment by sandeep-ps** [comment]
Hi, Yongwook, I'm reviewing this. Due to some reason, models.subprogram contains Subprogram instead of SubProgram (which was in the CRP branch, and hence now in the develop branch). Currently, there are no merge conflicts with the develop branch. Due to this, the ACEP and CRP endpoints and failing in local. I'm fixing it in this branch. I hope that's okay.

**Comment by ywkim312** [comment]
> Hi, Yongwook, I'm reviewing this. Due to some reason, models.subprogram contains Subprogram instead of SubProgram (which was in the CRP branch, and hence now in the develop branch). Currently, there are no merge conflicts with the develop branch. Due to this, the ACEP and CRP endpoints and failing in local. I'm fixing it in this branch. I hope that's okay.

Sure, no problem. Thanks for your help! I really appreciate it.

File: app/controllers/pdl.py [hunk 1]
```diff
@@ -660,32 +662,34 @@ def titles_title_ii_programs_rcpp_summary_search():
...
-        data_json = json.loads(file_data, object_pairs_hook=OrderedDict)
-
-        return data_json
+    program_id = get_program_id(TTTLE_IX_CROP_INSURANCE_PROGRAM_NAME)
+    if program_id is None:
+        msg = {
+            "reason": "No record for the given program name " + TTTLE_IX_CROP_INSURANCE_PROGRAM_NAME,
+            "error": "Not found: " + request.url,
+        }
+        logging.error("CRP: " + json.dumps(msg))
```
Comment (sandeep-ps) [inline app/controllers/pdl.py:671]:
```suggestion
        logging.error("Crop Insurance: " + json.dumps(msg))
```

File: app/controllers/pdl.py [hunk 2]
```diff
@@ -660,32 +662,34 @@ def titles_title_ii_programs_rcpp_summary_search():
...
```
Comment (sandeep-ps) [inline app/controllers/pdl.py:687]:
```suggestion
        logging.error("Crop Insurance: " + json.dumps(msg))
```

File: app/controllers/pdl.py [hunk 3]
```diff
@@ -88,6 +88,8 @@
 TITLE_II_ACEP_PROGRAM_NAME = "Agricultural Conservation Easement Program (ACEP)"
 TITLE_II_RCPP_PROGRAM_NAME = "Regional Conservation Partnership Program (RCPP)"
 TITLE_IV_SNAP_PROGRAM_NAME = "Supplemental Nutrition Assistance Program (SNAP)"
+TTTLE_IX_CROP_INSURANCE_PROGRAM_NAME = "Crop Insurance"
```
Comment (sandeep-ps) [inline app/controllers/pdl.py:91]:
```suggestion
TTTLE_XI_CROP_INSURANCE_PROGRAM_NAME = "Crop Insurance"
```
Fixing a minor typo in the Title number. Please update the variable name in this file.

File: app/models/subprogram.py [hunk 4]
```diff
@@ -14,7 +14,7 @@ def __init__(self, program_id, name):
         self.name = name
 
     def __repr__(self):
-        return 'SubPrograms(program_id=%s, name=%s)' % (self.program_id, self.name)
+        return 'Subprograms(program_id=%s, name=%s)' % (self.program_id, self.name)
```
Comment (sandeep-ps) [inline app/models/subprogram.py:17]:
```suggestion
        return 'SubPrograms(program_id=%s, name=%s)' % (self.program_id, self.name)
```

File: app/controllers/pdl.py (hunk 1 above)
Comment (ywkim312) [reply app/controllers/pdl.py:671]: Changed. My bad.

File: app/controllers/pdl.py (hunk 2 above)
Comment (ywkim312) [reply app/controllers/pdl.py]: Changed. Thanks for finding out.

File: app/controllers/pdl.py (hunk 3 above)
Comment (ywkim312) [reply app/controllers/pdl.py:91]: Changed. Nice catch, thanks

File: app/models/subprogram.py (hunk 4 above)
Comment (ywkim312) [reply app/models/subprogram.py]: Changed

File: app/controllers/pdl.py (hunk 3 above)
Comment (sandeep-ps) [reply app/controllers/pdl.py:91]: Hi, Yongwook, it looks like this was not changed.

Comment (ywkim312) [reply app/controllers/pdl.py:91]: changed

**Review by sandeep-ps — APPROVED** [review]
Approving. Thanks!

**Pull request merged**